### ISSUE TITLE
ExpectedException.expectMessage AssertionError lost with PowerMockRunner

### ISSUE THREAD

**Issue by johanhaleby** [jayway/powermock#396]
_From [dietrich...@gmail.com](https://code.google.com/u/115844465204731828022/) on March 09, 2012 07:58:14_

What steps will reproduce the problem? Run this test class:

@RunWith(PowerMockRunner.class)
public class ExpectedExceptionTest {

	@Rule
	public ExpectedException thrown = ExpectedException.none();

	@Test
	public void testExpectedException()  {
		thrown.expect(IllegalStateException.class);
		thrown.expectMessage("expected message");

		throw new IllegalStateException("surprise message");
	}

} What is the expected output? What do you see instead? Expected output:
java.lang.AssertionError: 
Expected: (exception with message a string containing "expected message" and an instance of java.lang.IllegalStateException)
     got: \<java.lang.IllegalStateException: surprise message>

Instead I see:
java.lang.IllegalStateException: surprise message

Without the PowerMockRunner the expected AssertionError is shown in the output. What version of the product are you using? On what operating system? 1.4.11

_Original issue: http://code.google.com/p/powermock/issues/detail?id=376_

**Comment by johanhaleby** [comment]
_From [johan.ha...@gmail.com](https://code.google.com/u/105676376875942041029/) on March 08, 2012 23:36:15_

What version of Junit are you using?

**Comment by johanhaleby** [comment]
_From [ronald.b...@gmail.com](https://code.google.com/u/108393075678473797769/) on September 26, 2013 09:04:48_

I also experienced this when using PowerMockRule and ExpectedException rule. 
It seems the problem is the order of the Rule's statements being executed and the classloading PowerMock does. The exception is thrown OK through to ExpectedExceptionStatement, but the ExpectedExceptionStatement itself is a different object than the one that is visible to the test (I suspect due to classloading).
That means that the "expect" methods in the Test are working with another Object than the one that does the "handleException", and so there are no matchers and the whole thing fails.

A workaround is as follows:
1.) create a class PowerMockTestRule which wraps PowerMockRule to make it implement TestRule:

import java.lang.reflect.Method;

import org.junit.rules.TestRule;
import org.junit.runner.Description;
import org.junit.runners.model.FrameworkMethod;
import org.junit.runners.model.Statement;
import org.powermock.modules.junit4.rule. PowerMockRule ;

public class PowerMockTestRule implements TestRule {

	private final PowerMockRule rule;
	public PowerMockTestRule( PowerMockRule rule) {
		this.rule = rule;
	}
	@Override
	public Statement apply(Statement base, Description description) {
		return rule.apply(base, createFrameworkMethod(description), getTestObject(description));
	}
	private FrameworkMethod createFrameworkMethod(Description description) {
		try {
			String methodName = description.getMethodName();
			Class c = getTestClass(description);
			Method m = c.getDeclaredMethod(methodName);
			return new FrameworkMethod(m);
		} catch (Exception e) {
			throw new IllegalStateException(e);
		}
	}
	private Class<?> getTestClass(Description description) {
		return description.getTestClass();
	}
	
	private Object getTestObject(Description description) {
		try {
			return getTestClass(description).newInstance();
		} catch (InstantiationException | IllegalAccessException e) {
			throw new IllegalStateException(e);
		}
	}

}


2.) Use RuleChain to define the order of the Rules:
public ExpectedException thrown = ExpectedException.none();
@Rule public TestRule chain = RuleChain
			.outerRule(new PowerMockTestRule(new PowerMockRule ()))
			.around(thrown);

I don't know if this is feasible for a general solution, but in my case, it works.


PS: Junit used: 4.11, Powermock used: 1.5.1

**Comment by johanhaleby** [comment]
_From [ronald.b...@gmail.com](https://code.google.com/u/108393075678473797769/) on September 26, 2013 13:02:45_

Update: instead of the Powermock specific PowerMockTestRule above i am now using a more generic Version: TestRuleAdapter:

import java.lang.reflect.Method;

import org.junit.rules.MethodRule;
import org.junit.rules.TestRule;
import org.junit.runner.Description;
import org.junit.runners.model.FrameworkMethod;
import org.junit.runners.model.Statement;

public class TestRuleAdapter implements TestRule {

	private final MethodRule rule;
	public TestRuleAdapter(MethodRule rule) {
		this.rule = rule;
	}
	@Override
	public Statement apply(Statement base, Description description) {
		return rule.apply(base, createFrameworkMethod(description), getTestObject(description));
	}
	private FrameworkMethod createFrameworkMethod(Description description) {
		try {
			String methodName = description.getMethodName();
			Class<?> c = getTestClass(description);
			Method m = c.getDeclaredMethod(methodName);
			return new FrameworkMethod(m);
		} catch (Exception e) {
			throw new IllegalStateException(e);
		}
	}
	private Class<?> getTestClass(Description description) {
		return description.getTestClass();
	}
	
	private Object getTestObject(Description description) {
		try {
			return getTestClass(description).newInstance();
		} catch (InstantiationException | IllegalAccessException e) {
			throw new IllegalStateException(e);
		}
	}

}

**Comment by johanhaleby** [comment]
_From [johan.ha...@gmail.com](https://code.google.com/u/105676376875942041029/) on September 26, 2013 22:16:35_

Thanks for sharing this. It wouldn't be possible for you to try to resolve this issue in the PowerMockRule ?  I would be really glad if we could get this fixed to the next release.

**Comment by johanhaleby** [comment]
_From [kyk...@gmail.com](https://code.google.com/u/100724633671938180861/) on February 05, 2015 23:23:38_

Is this expected to be fixed anytime soon? I am on JUnit 4.12, PowerMock 1.6.1, and Mockito 1.10.8 and this still doesn't work. Anytime I use PowerMock on a test class I have to revert to try/catch to validate exceptions.

**Comment by johanhaleby** [comment]
_From [ryan.dan...@gmail.com](https://code.google.com/u/100684921030866931749/) on March 24, 2015 15:03:38_

This is a serious problem. To fix it, all that needs to be done is for PowerMockRule to implement TestRule so that it can be chainable with RuleChain.

It is currently not possible to submit issues on Github for this that I could see.

**Comment by johanhaleby** [comment]
_From [johan.ha...@gmail.com](https://code.google.com/u/105676376875942041029/) on March 26, 2015 12:55:41_

No issues are currently reported here (but we will move to Github). Please submit a pull request at Github and I'll include it.

**Comment by maratfdrv** [comment]
It still reproduced in junit 4.12 and powermock 1.6.6

**Comment by thekingnothing** [comment]
@maratfdrv, thank you for raising an issue. As I understand, pull request was not created and bug is still presence in code. I've added to backlog and I hope it will be fixed in next release.

**Comment by rockydcoder** [comment]
@thekingnothing, can I take this up? I can see two `PowerMockRule` classes. One in `powermock-module-junit4-rule` and one in `powermock-module-junit4-rule-agent`

**Comment by thekingnothing** [comment]
@rockydcoder yep, you can. There are two `PowerMockRule`: one for case when PowerMock reloads already loaded classes with PowerMock class loader and second when Java Agent is used to instrument classes.

**Comment by thekingnothing** [comment]
By the way, there is such test as `AssertPowerMockRuleDelagatesToOtherRulesTest` you may modify it to check that exception is not lost.

**Comment by rockydcoder** [comment]
Hmm.. I do not think this is as simple as it looks. If `PoserMockRule` implements `TestRule` and overrides `apply` method as discussed above, I get a `com.thoughtworks.xstream.mapper.CannotResolveClassException` if I try to run the sample test.

**Comment by rockydcoder** [comment]
Ignore the previous comment. 
I was able to resolve the bug but a `RuleChain` is still needed to define the order. Should that be classified as a bug too?

**Comment by rockydcoder** [comment]
@thekingnothing any updates on this?

**Comment by thekingnothing** [comment]
@rockydcoder , sorry for delay with replay.  As I understand you, issue could be fixed and rules could work together only in case if `RuleChain` define and there is no way to implement `PowerMockRule` to work independently without `RuleChain`. Am I right? Is so, then we should point it in documentation on wiki page and not create a new bug.

**Comment by rockydcoder** [comment]
@thekingnothing I created a PR but it fails in Travis. I see there are two failed tests but I am unable to run them on my local machine. I am not aware of Java Agent thingy. Can you point me to some resources on the web?

**Comment by thekingnothing** [comment]
You may find more information about Powermock and JavaAgent here
https://github.com/powermock/powermock/wiki/PowerMockAgent


Best regrads,
Arthur Zagretdinov


On 26 Mar 2017, at 17:00, Rakesh Sarangi <notifications@github.com<mailto:notifications@github.com>> wrote:


@thekingnothing<https://github.com/thekingnothing> I created a PR but it fails in Travis. I see there are two failed tests but I am unable to run them on my local machine. I am not aware of Java Agent thingy. Can you point me to some resources on the web?

**Comment by rockydcoder** [comment]
From what I understand, it is not wise for a rule to implement both `MethodRule` and `TestRule`, as has been said [here](https://github.com/junit-team/junit4/issues/351#issuecomment-102084524)
Any comments? @thekingnothing
